Dashboard View

jvanderaa opened the pull request:
WIP yet, this is the work towards building a Dashboard view for the components of the circuit maintenance.

Outstanding:
- [x] Complete planned dashboard metrics
- [ ] Integrate functions/methods into Capacity Metrics

glennmatthews commented on nautobot_circuit_maintenance/urls.py:
```diff
@@ -6,6 +6,9 @@
 urlpatterns = [
+    # Overview
+    # path("config-compliance/overview/", views.ConfigComplianceOverview.as_view(), name="configcompliance_report"),
+    path("maintenance/overview/", views.CircuitMaintenanceOverview.as_view(), name="circuitmaintenance_overview"),
```
Might be good to register this as the "home" page for the plugin? https://nautobot.readthedocs.io/en/stable/plugins/development/#adding-links-to-the-installed-plugins-view

jvanderaa replied:
Making that update. What does having a home page do for a plugin?

glennmatthews replied:
It gets linked from the `installed-plugins` list page, i.e.:

![image](https://user-images.githubusercontent.com/5603551/178288271-7d4edd38-7145-493d-8243-c3645525dac2.png)

jvanderaa replied:
Cool. I had a feeling that was the case. What awesome stuff!

scetron commented:
I think this is a great addition to the plugin and agree with @glennmatthews  that it would make a good plugin landing/home page.

jvanderaa commented on pyproject.toml:
```diff
@@ -92,12 +94,20 @@ no-docstring-rgx="^(_|test_|Meta$)"
 # Pylint and Black disagree about how to format multi-line arrays; Black wins.
+#
+# agruments-differ:
+# 21:4: W0221: Number of parameters was 1 in 'TestCase.setUp' and is now 1 in overridden 'GraphQLTestCase.setUp'
```
This does look like a change is needed. The BaseForm class changed to include an item for approvals.

glennmatthews replied:
See https://github.com/nautobot/nautobot-plugin-circuit-maintenance/pull/195#discussion_r918920298

jvanderaa commented:
I'm currently waiting on #195 before I continue on.

chadell commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -1,20 +1,197 @@
+            "Average Duration of Maintenances": average_maintenance_duration,
+            "Future Maintenances": upcoming_maintenance_count,
+            "Average Number of Maintenances Per Month": round(self.get_maintenances_per_month(), 1),
+            "Next 30 Days, Maintenance to Circuit Ratio": round(
```
this is not always 30, it depends on user configuration

chadell commented on nautobot_circuit_maintenance/__init__.py:
```diff
@@ -89,8 +89,10 @@ class CircuitMaintenanceConfig(PluginConfig):
     default_settings = {
         "raw_notification_initial_days_since": 7,
         "raw_notification_size": 8192,
+        "dashboard_n_days": 30,
```
we should docuement this option in the readme

chadell commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -1,20 +1,197 @@
 
 
+class CircuitMaintenanceOverview(generic.ObjectListView):
+    """View for an overview dashboard of summary view."""
```
I would appreciate a summary of the info provided in this overview (the report keys)

jvanderaa replied:
Help me out on this one, are we talking what we are passing into the templates?

chadell replied:
I was thinking on a summary of the information that this view is providing, in a bullet point list. So, then you explore the code with some context.

jvanderaa replied:
Shoot, I need to do this yet.

chadell commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -1,20 +1,197 @@
+        return count
+
+    @staticmethod
+    def get_month_list():
```
adding some basic unittesting for all these functions would be great

jvanderaa replied:
I agree. I need to put some brain power into this one again. It has just been a bit too long for me at the moment. It will probably be better than the manual testing that I have been doing then too.

jvanderaa replied:
I need to add a few more tests tomorrow, including the next couple that we see below. I'm excited to do some more of these.

chadell reviewed (changes requested):
Calling the `overview` view, without any Circuit Maintenance, raises an `AttributeError` exception

```
AttributeError at /plugins/circuit-maintenance/maintenance/overview/
'NoneType' object has no attribute 'start_time'
Request Method:	GET
Request URL:	http://localhost:8080/plugins/circuit-maintenance/maintenance/overview/
Django Version:	3.2.15
Exception Type:	AttributeError
Exception Value:	
'NoneType' object has no attribute 'start_time'
Exception Location:	/source/nautobot_circuit_maintenance/views.py, line 169, in get_month_list
Python Executable:	/usr/local/bin/python
Python Version:	3.7.13
Python Path:	
['/source',
 '/usr/local/lib/python3.7/site-packages/git/ext/gitdb',
 '/source',
 '/usr/local/bin',
 '/usr/local/lib/python37.zip',
 '/usr/local/lib/python3.7',
 '/usr/local/lib/python3.7/lib-dynload',
 '/root/.local/lib/python3.7/site-packages',
 '/usr/local/lib/python3.7/site-packages',
 '/source',
 '/usr/local/lib/python3.7/site-packages/gitdb/ext/smmap']
Server time:	Wed, 24 Aug 2022 07:08:07 +0000
```

chadell reviewed (changes requested):
Accessing the `overview` view with a `CircuitMaintenance` without circuits, trigger a `ZeroDivisionError` exception

```
ZeroDivisionError at /plugins/circuit-maintenance/maintenance/overview/
division by zero
Request Method:	GET
Request URL:	http://localhost:8080/plugins/circuit-maintenance/maintenance/overview/
Django Version:	3.2.15
Exception Type:	ZeroDivisionError
Exception Value:	
division by zero
Exception Location:	/source/nautobot_circuit_maintenance/views.py, line 69, in setup
Python Executable:	/usr/local/bin/python
Python Version:	3.7.13
Python Path:	
['/source',
 '/usr/local/lib/python3.7/site-packages/git/ext/gitdb',
 '/source',
 '/usr/local/bin',
 '/usr/local/lib/python37.zip',
 '/usr/local/lib/python3.7',
 '/usr/local/lib/python3.7/lib-dynload',
 '/root/.local/lib/python3.7/site-packages',
 '/usr/local/lib/python3.7/site-packages',
 '/source',
 '/usr/local/lib/python3.7/site-packages/gitdb/ext/smmap']
Server time:	Wed, 24 Aug 2022 07:09:54 +0000
```

chadell reviewed (changes requested):
I create a CM for the next 30 Days, and is not showing up in the dashboard (this circuit is not associated to a Device)

<img width="1172" alt="image" src="https://user-images.githubusercontent.com/15998111/186354766-cb8eb808-a1ca-440f-a79a-d157704093be.png">

<img width="1395" alt="image" src="https://user-images.githubusercontent.com/15998111/186354824-f0517478-148c-4cc3-af0b-3a43f3f7e500.png">

chadell commented on nautobot_circuit_maintenance/navigation.py:
```diff
@@ -3,6 +3,11 @@
 menu_items = (
+    PluginMenuItem(
+        link="plugins:nautobot_circuit_maintenance:circuitmaintenance_overview",
+        link_text="Circuit Maintenance Dashboard",
```
what about?

```suggestion
        link_text="Dashboard",
```

jvanderaa replied:
`CM Dashboard` work to put a small specific to it?

chadell replied:
being within the "context" of "Circuit Maintenance", it seems redundant to me. Why do you think is relevant to be specific about it?

jvanderaa replied:
In my design preferences I like to see the name of it to be a bit more explicit about what dashboard I'm going to. Let's say there are 3 plugins inside of circuits, and all of them have "Dashboard" for an option. This can add confusion about which dashboard in particular it is that is being referred to if you miss the heading section.

jvanderaa replied:
Updated regardless.

chadell replied:
I'm not strong on this point, just trying to get more inputs

jvanderaa commented:
> I create a CM for the next 30 Days, and is not showing up in the dashboard (this circuit is not associated to a Device)
> 
> <img alt="image" width="1172" src="https://user-images.githubusercontent.com/15998111/186354766-cb8eb808-a1ca-440f-a79a-d157704093be.png">
> 
> <img alt="image" width="1395" src="https://user-images.githubusercontent.com/15998111/186354824-f0517478-148c-4cc3-af0b-3a43f3f7e500.png">

This one looks like it is a caching issue. I was able to replicate it in my environment. I was able to get it to show up by canceling out of the development environment and restarting it. Showed up immediately.

jvanderaa commented:
Resolved the division by zero as well, it stemmed from a query with 0 results in the list.

jvanderaa commented:
> Accessing the `overview` view with a `CircuitMaintenance` without circuits, trigger a `ZeroDivisionError` exception
> 
> ```
> ZeroDivisionError at /plugins/circuit-maintenance/maintenance/overview/
> division by zero
> Request Method:	GET
> Request URL:	http://localhost:8080/plugins/circuit-maintenance/maintenance/overview/
> Django Version:	3.2.15
> Exception Type:	ZeroDivisionError
> Exception Value:	
> division by zero
> Exception Location:	/source/nautobot_circuit_maintenance/views.py, line 69, in setup
> Python Executable:	/usr/local/bin/python
> Python Version:	3.7.13
> Python Path:	
> ['/source',
>  '/usr/local/lib/python3.7/site-packages/git/ext/gitdb',
>  '/source',
>  '/usr/local/bin',
>  '/usr/local/lib/python37.zip',
>  '/usr/local/lib/python3.7',
>  '/usr/local/lib/python3.7/lib-dynload',
>  '/root/.local/lib/python3.7/site-packages',
>  '/usr/local/lib/python3.7/site-packages',
>  '/source',
>  '/usr/local/lib/python3.7/site-packages/gitdb/ext/smmap']
> Server time:	Wed, 24 Aug 2022 07:09:54 +0000
> ```

Resolved.

chadell commented on pyproject.toml:
```diff
@@ -90,12 +90,20 @@ no-docstring-rgx="^(_|test_|Meta$)"
 # Line length is enforced by Black, so pylint doesn't need to check it.
 # Pylint and Black disagree about how to format multi-line arrays; Black wins.
+#
+# agruments-differ:
```
```suggestion
# arguments-differ:
```

chadell reviewed (commented):
"Average Duration of Maintenances" is always 0? I added some future circuit maintenances, and always show as 0

jvanderaa commented:
> "Average Duration of Maintenances" is always 0? I added some future circuit maintenances, and always show as 0

Corrected. The `seconds` property was initially giving good results. But the proper method to use was `total_seconds()` to get the duration within seconds on a timedelta object.

glennmatthews commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -1,20 +1,226 @@
+            int: Count of future maintenances
+        """
+        count = 0
+        for ckt_maint in self.queryset.all():
+            if ckt_maint.start_time.date() > start_date:
+                count += 1
```
```python
count = self.queryset.filter(start_time__gte=start_date_midnight).count()
```

glennmatthews commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -1,20 +1,226 @@
+        Returns:
+            float: Average maintenances per month
+        """
+        # Initialize each month of maintenances
+        months = self.get_month_list()
```
this feels like a lot more processing and work around the months than needed. Why not just something like:

```python
ordered_ckt_maintenance = CircuitMaintenance.objects.order_by("start_time")
if ordered_ckt_maintenance.count() < 2:
    return 0
delta = ordered_ckt_maintenance.last().start_time - ordered_ckt_maintenance.first().start_time
delta_months = delta.days / 365.0 * 12.0
if delta_months == 0:
    return 0
return self.queryset.count() / delta_months
```

jvanderaa replied:
I have to investigate some more on this. This counts the days in between rather than the month itself. If there were only the months in the given year, this wouldn't be too bad. But then we get off a bit. The calculation gives a slightly off on the unit test.

jvanderaa replied:
Found a way that will do this! Incoming.

jvanderaa commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/circuit_maintenance_overview.html:
```diff
@@ -0,0 +1,135 @@
+            <h3 class="text s-2 p-3">Circuits With Maintenances in the Next {{ n_days }} Days</h3>
+            <table class="table table-hover table-headings">
+                <thead>
+                    <tr class="even">
```
Stolen from the GC plugin. Will attempt to remove it.

jvanderaa replied:
Removal successful. ✅

jvanderaa commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/circuit_maintenance_overview.html:
```diff
@@ -0,0 +1,135 @@
+                            <td>
+                                {% if circuit.termination_a %}
+                                    {% with circuit_term_a_site=circuit.termination_a.connected_endpoint.parent.site %}
+                                <a href="{{ circuit_term_a_site.get_absolute_url }}">{{ circuit_term_a_site.name }}</a>{% if circuit.termination_z.connected_endpoint and circuit_term_a_site != circuit.termination_z.connected_endpoint.parent.site %}, <a href="{{ circuit.termination_z.connected_endpoint.parent.site.get_absolute_url }}">{{ circuit.termination_z.connected_endpoint.parent.site.name}}< ...
```
Any suggestions? If I add a newline in the template I get an unnatural space in the text.

jvanderaa commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/circuit_maintenance_overview.html:
```diff
@@ -0,0 +1,135 @@
+                                {{ circuit.status.name }}
+                            </td>
+                            <td>
+                                {{ maintenance.start_time | date:'Y-m-d H:i' }}
```
✅

jvanderaa commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/circuit_maintenance_overview.html:
```diff
@@ -0,0 +1,135 @@
+                <tbody>
+                    {% for maintenance in upcoming_maintenances %}
+                    {% for circuit in maintenance.circuits %}
+                    {% if circuit.termination_a %}
```
✅

jvanderaa commented on nautobot_circuit_maintenance/tests/test_views.py:
```diff
@@ -407,3 +436,323 @@ def test_has_advanced_tab(self):
+        return f"plugins:{self.model._meta.app_label}:{self.model._meta.model_name}_{{}}"
+
+    def assertInstanceEqual(self, instance, data, api=False):  # pylint: disable=arguments-differ
+        """Used to overwrite inbuilt function. Causing type issues for datetimepicker."""
```
@chadell thoughts here? I believe that I used this from an override that you had added.

chadell replied:
This was added 16months ago to some issue with the datetimepicker comparison. However, maybe since then, this issue could has been solved and no longer needed. 
does it ring some bell @glennmatthews ?

jvanderaa commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -1,20 +1,226 @@
+    template_name = "nautobot_circuit_maintenance/circuit_maintenance_overview.html"
+    queryset = models.CircuitMaintenance.objects.all()  # Needs to remain all objects, otherwise other calcs will fail.
+
+    def setup(self, request, *args, **kwargs):  # pylint: disable=too-many-locals
```
I'll love to learn more on that. Moved and working.

jvanderaa commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -1,20 +1,226 @@
+            Set: Set of maintenances that are up coming
+        """
+        end_date = start_date + datetime.timedelta(days=n_days)
+        maintenances = self.queryset.all()
+        return_list = []
+        for maintenance in maintenances:
+            if start_date <= maintenance.start_time.date() <= end_date:
+                return_list.append(maintenance)
```
Looks better.

chadell commented on nautobot_circuit_maintenance/tests/test_views.py:
```diff
@@ -407,3 +436,323 @@ def test_has_advanced_tab(self):
         pass
+
+
+class DashboardTest(ViewTestCases.PrimaryObjectViewTestCase):
```
I was going to say the same...

jvanderaa commented on nautobot_circuit_maintenance/views.py:
```diff
@@ -180,45 +170,26 @@ def total_months(datetime_obj):
-        # This will handle any Division by 0 errors
-        if len(months) == 0:
+        if delta_months == 0:
             return 0
```
✅

chadell reviewed (commented):
It's almost a 👍 , I would only simply the testing of the new view (as suggested by Glenn) with a simpler test class, without all the tests that apply to a model, but in this case it's only for an extra view. It should get rid of a lot of the `skip` that are not adding any value

jvanderaa commented:
Done @chadell

chadell reviewed (dismissed):
LGTM
🎉 Amazing work Josh!
I see this just a first iteration of the dashboard, where we could extend and add useful information as we add more features.

Regarding testing, for the `assertInstanceEqual` weirdness, I would create a housekeeping issue to fix it.

chadell commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/circuit_maintenance_overview.html:
```diff
@@ -0,0 +1,129 @@
+{% table_config_form table table_name="ObjectTable" %}
+{% endblock %}
+{% block javascript %}
+<script src="{% static 'js/tableconfig.js' %}"></script>
```
@jvanderaa , could you give explain (point to doc) what are we using this for?

jvanderaa replied:
@itdependsnetworks I took this from the GC plugin. Any insight?

jvanderaa replied:
Removed it for now. It is used for table selection.

jvanderaa replied:
Let's get this merged and released?

glennmatthews replied:
It's used for customizing a table view to specify which columns are displayed. If you're removing it, you should also remove the `table_config_form` template tag as well.

scetron reviewed (dismissed):
Looks good to me

chadell commented on nautobot_circuit_maintenance/templates/nautobot_circuit_maintenance/circuit_maintenance_overview.html:
```diff
@@ -0,0 +1,126 @@
+    </div>
+</div>
+
+{% table_config_form table table_name="ObjectTable" %}
```
@jvanderaa  From Glenn's comment.
```suggestion
```

... and let's get in and released!

chadell reviewed (approved):
LGTM
